<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>API</title>
</head>
<body>
  
</body>
<script>
  // new Promise((resolv,reject) => {
  //   setTimeout(() => {
  //     resolv('成功的数据');
  //     // reject('失败的数据');
  //   },1000)
  // }).then((value ) => {
  //   console.log('onResolved()' + value);
  // }).catch((reason) => {
  //   console.log('onRejected()' + reason);
  // })

  // // 1、Promise函数对象直接调用成功|失败方法并传递值
  // const p1 = Promise.resolve(1);
  // const p3 = Promise.resolve(3);
  // const p2 = Promise.reject(2);
  // p1.then(value => {console.log(value)});
  // p2.catch(reason => {console.log(reason)});

  // //2、Promise.all([]) 接收一个数组，当数组中有回调失败则执行catch返回回调失败的内容，反之则使用数组返回回调成功的内容
  // const pall = Promise.all([p1,p3]);
  // // const pall = Promise.all([p1,p2,p3]);
  // pall.then(value => {
  //   console.log('onResolved()',value);  //onResolved() (2) [1, 3]
  // }).catch(reason => {
  //   console.log('onRejected()',reason); //6promiseApi.html:36 onRejected() 2
  // })

  // //3、Promise.race([]) 接收一个数组，看数组中那个先执行完成，就返回那个，当第一个执行完成后，其他的就不用执行了
  // const pra = Promise.race([p1,p2,p3]);
  // pra.then(value => {
  //   console.log('onResolved()',value);
  // }).catch(reason => {
  //   console.log('onRejected()',reason);
  // })

  // ===================================================================
  // new Promise((resolve, reject) => {
  //   setTimeout(() => {
  //     resolve('success')
  //   }, 1000)
  // })
  // .then(value => {
  //   console.log(value)
  //   return Promise.reject(new Error('error'))
  // })
  // .catch(reason => {
  //   console.log(reason.message)
  // })

  // 产生一个成功值为 1 的promise对象
  const p1 = new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve('1')
    })
  })

  // p1.then(value => {
  //   console.log(value)
  // })
  const p2 = Promise.resolve(2)
  const p3 = Promise.reject(3)
  
  // const pAll = Promise.all([p1, p2, p3]).then(value => {
  //   console.log(value)
  // }).catch(reason => {
  //   console.log(reason)
  // })

  const pReac = Promise.race([p1, p2, p3]).then(value => {
    console.log(`value: ${value}`)
  }).catch(reason => {
    console.log(`reason: ${reason}`)
  })
</script>
</html>